src/cargo/sources/git/oxide.rs

```diff
@@ -49,22 +49,78 @@ pub fn with_retry_and_progress(
         .ok()
         .unwrap_or_default()
         .auto_deregister();
-    let should_interrupt = AtomicBool::new(false);
     let mut progress_bar = Progress::new("Fetch", config);
     std::thread::scope(move |s| {
         s.spawn({
             let root = Arc::downgrade(&progress_root);
             move || -> CargoResult<()> {
+                const READ_PACK_BYTES: [u8; 4] = *b"BWRB";
+                const DELTA_INDEX_OBJECTS: [u8; 4] = *b"IWIO";
+                const RESOLVE_OBJECTS: [u8; 4] = *b"IWRO";
+
+                // We choose `N=10` here to make a `300ms * 10slots ~= 3000ms`
+                // sliding window for tracking the data transfer rate (in bytes/s).
+                let mut last_update = Instant::now();
+                let mut counter = MetricsCounter::<10>::new(0, last_update);
+
                 let mut tasks = Vec::with_capacity(10);
                 while let Some(root) = root.upgrade() {
                     root.sorted_snapshot(&mut tasks);
-                    progress_bar.tick(0, 10, "TBD")?;
-                    // dbg!(&tasks);
+                    let counters = tasks
+                        .iter()
+                        .find_map(|(_, t)| {
+                            (t.id == READ_PACK_BYTES).then(|| t.progress.as_ref().expect("set"))
```

Comment:
It's indeed very rough 😅, thanks for surfacing this atrocity.

The intention is to make these IDs part of the API and its stability guarantees, and for that to happen they would have to be exposed by `gitoxide` crates themselves to improve the situation and not have to know these values in the code of the consumer.

Something that probably can't be improved is that these IDs can only be associated with a method setting them _by convention_, as there is no way to associate them with the method itself. Maybe a good middle ground would be to use associated constants so they are at least associated with the type containing the method, and be in the same impl block.

I will experiment with that and at least get these IDs out of `cargo`.

Reply:
Even with constants, I have the feeling that this is exposing the wrong level of abstraction if we are dealing with `[u8; 4]` types for ids, it gives me the impression when reading this that we are dealing with raw data.

I feel like I should be dealing with an enum instead.  The next level would likely be a newtype that *might* poke the abstraction to allow access to the byte strings.

To be clear, I say all of this not having experience with what this part is doing.  My experience is with git2 and I've not touched progress work.  There very well could be valid reasons for all of this.

Reply:
❤️ Lovely!

Exposing raw bytes (even if it's behind a constant) name is worse than having an enum for it, as enums allow to limit the set of possible variants which is exactly what we want to express. That way, IDs can be an implementation detail up to the point where they are used to find the right progress bar.

I think this is as far as it can go, and I see no way to hide them more. In any case, the TODO list was updated.